sound/soc/sof/topology.c

```diff
@@ -1781,6 +1924,39 @@ static int sof_widget_load_process(struct snd_soc_component *scomp, int index,
 	return 0;
 }
 
+static int sof_widget_bind_event(struct snd_sof_dev *sdev,
+				 struct snd_sof_widget *swidget,
+				 u16 event_type)
+{
+	struct sof_ipc_comp *ipc_comp;
+
+	/* validate widget event type */
+	switch (event_type) {
+	case SOF_KEYWORD_DETECT_DAPM_EVENT:
+		/* only KEYWORD_DETECT comps should handle this */
+		if (swidget->id != snd_soc_dapm_effect)
+			goto err;
+
+		ipc_comp = swidget->private;
+		if (ipc_comp && ipc_comp->type != SOF_COMP_KEYWORD_DETECT)
+			goto err;
+		break;
+	default:
+		goto err;
+	}
+
+	/* bind event to widget */
+	return snd_soc_tplg_widget_bind_event(swidget->widget,
+					      sof_tplg_widget_ops,
+					      ARRAY_SIZE(sof_tplg_widget_ops),
+					      event_type);
```

Comment:
you can simplify the flow by moving this bind_event() call inside of the switch, and renaming the widget_ops and keyword_widget_ops.

Reply:
@plbossart updated now.